File: hw2/Javascript/Part2/part2.js
// Data structures: Array, list, dictionary, sorted list, hashset, sortedset, queue, stack, linkedlist

// Array
let arr = [0,10,22,31,441];
for (let i = 0; i < arr.length; i++) {
    if (arr[i] == 22){
        break;
    }
}

for (let i = 0; i < arr.length; i++) {
    if (arr[i] == 31){
        console.log(arr[i]);
        continue;
    }
}
let element = 100;
arr.push(element); // Add element
let index = 3;
arr.splice(index,1); // Remove one element at index 3 
arr[index-1]; // Get element by index
arr.includes(element); // Check existence of element

// List (simulated using Array in JavaScript)
// Same operations as Array

// Dictionary (JavaScript object)
const obj = {};
for (const key in obj) {
    if (key == someValue){
        break;
    }else{
        console.log(key);
        continue;
    }
}
let key = 42;
obj[key] = value; // Add key/value
let value = 18;
delete obj[key]; // Remove key/value
obj[key]; // Get value by key
obj.hasOwnProperty(key); // Check existence

// Sorted List (sort an Array in JavaScript)
// Same operations as Array, but maintain sorted order after add

// HashSet (using JavaScript Set)
const set = new Set();
element = 200;
for (const el in set) {
    if (el){
        //Do something
        break;
    }
    if (el == element){
        console.log(el);
        continue;
    }
}

set.add(element); // Add element
set.delete(element); // Remove element
set.has(element); // Check existence

// SortedSet (using an Array and sorting)
const sortedSet = [];
element = 13;
for (const elOrd in sortedSet) {
    if (elOrd){
        //Do something
        break;
    }
    if (elOrd == element){
        console.log(elOrd);
        continue;
    }
}

// Add element (and maintain sorted order)
sortedSet.push(element);
sortedSet.sort();
sortedSet.includes(element); // Check existence

// Queue (using Array)
const queue = [];
element = 109;
for (const el in queue) {
    if (el){
        //Do something
        break;
    }
    if (el == element){
        console.log(el);
        continue;
    }
}
queue.push(element); // Enqueue
queue.shift(); // Dequeue
queue[0]; // Get front element
queue.length > 0; // Check non-empty

// Stack (using Array)
const stack = [];
element = 11;
for (const stackEl in stack) {
    if (stackEl){
        //Do something
        break;
    }
    if (stackEl == element){
        console.log(stackEl);
        continue;
    }
}
stack.push(element); // Push
stack.pop(); // Pop
stack[stack.length - 1]; // Get top element
stack.length > 0; // Check non-empty

// LinkedList (custom implementation)
let linkedList = new LinkedList();
  
// add elements
linkedList.append(1);
linkedList.append(6);
linkedList.append(0);
 
// loop
linkedList.print();
 
// remove elements
linkedList.removeElement(0);
linkedList.removeElement(6);
  
// check existance
linkedList.exists(0);
 
// retrieve elements
console.log(linkedList.retrieve(2))

File: hw2/C#/part2.cs
using System;
using System.Collections.Generic;

class Program
{
    static void Main()
    {
        // Data structures: Array, List, Dictionary, Sorted List, HashSet, SortedSet, Queue, Stack, LinkedList

        // Array
        int[] arr = new int[10];
        for (int i = 0; i < arr.Length; i++)
        {
            if (arr[i] == 5)
            {
                Console.WriteLine("Element found! Breaking the loop.");
                break; 
            }
        }

        Array.Resize(ref arr, arr.Length + 1);
        arr[arr.Length - 1] = element; // Add element
        Array.Clear(arr, index, 1); // Remove element by index
        int element = arr[index]; // Get element by index
        bool contains = Array.Exists(arr, e => e == element); // Check existence

        // List
        List<int> list = new List<int>();
        foreach (var item in list)
        {
            if (item == 7)
            {
                Console.WriteLine("Item found! Continue the loop.");
                continue; 
            }
        }

        list.Add(element); // Add element
        list.RemoveAt(index); // Remove element by index
        int element = list[index]; // Get element by index
        bool contains = list.Contains(element); // Check existence

        // Dictionary
        Dictionary<string, int> dict = new Dictionary<string, int>();
        foreach (var key in dict.Keys)
        {
            // Loop (break/continue) the same as list and array
        }

        dict[key] = value; // Add key/value
        dict.Remove(key); // Remove key/value
        bool containsKey = dict.ContainsKey(key); // Check existence of key
        bool containsValue = dict.ContainsValue(value); // Check existence of value

        // Sorted List (not available in C#, use SortedDictionary)
        // Use SortedDictionary<TKey, TValue>

        // HashSet
        HashSet<int> hashSet = new HashSet<int>();
        foreach (var item in hashSet)
        {
            // Loop (break/continue) the same as list and array
        }

        hashSet.Add(element); // Add element
        hashSet.Remove(element); // Remove element
        bool contains = hashSet.Contains(element); // Check existence

        // SortedSet
        SortedSet<int> sortedSet = new SortedSet<int>();
        foreach (var item in sortedSet)
        {
            // Loop (break/continue) the same as list and array
        }

        sortedSet.Add(element); // Add element
        sortedSet.Remove(element); // Remove element
        bool contains = sortedSet.Contains(element); // Check existence

        // Queue
        Queue<int> queue = new Queue<int>();
        foreach (var item in queue)
        {
            // Loop (break/continue) the same as list and array
        }

        queue.Enqueue(element); // Enqueue
        int front = queue.Dequeue(); // Dequeue
        int peek = queue.Peek(); // Peek front element
        bool nonEmpty = queue.Count > 0; // Check non-empty

        // Stack
        Stack<int> stack = new Stack<int>();
        foreach (var item in stack)
        {
            // Loop (break/continue) the same as list and array
        }

        stack.Push(element); // Push
        int pop = stack.Pop(); // Pop
        int peek = stack.Peek(); // Peek top element
        bool nonEmpty = stack.Count > 0; // Check non-empty

        // LinkedList
        LinkedList<int> linkedList = new LinkedList<int>();
        foreach (var item in linkedList)
        {
            // Loop (break/continue) the same as list and array
        }

        linkedList.AddLast(element); // Add element
        linkedList.Remove(element); // Remove element
        bool contains = linkedList.Contains(element); // Check existence
    }
}
